### PR TITLE
Adding MapFilter for transform and filter

### PR DESCRIPTION
**Pull request by calvinbrown085**
👋 I have a use case where i'd like to go from type A => B, but also filter in there. I figured a good way to achieve this was using a pointer type, where the map filters the nil type out.

Let me know what you think!

### CODE REVIEW

**Review by jha-ramesh — APPROVED** [review]
love it !! 

Do we want to have some kind of constraints in place for T ? (rather than `any` , would it be better to indicate the behavior from `T` (or type ) otherwise we would have to read `f` to understand the actual filter function)

**Comment by calvinbrown085** [comment]
@jha-ramesh 
> Do we want to have some kind of constraints in place for T ?

What kind of type constraints would you put on it?

**Comment by jha-ramesh** [comment]
> @jha-ramesh
> 
> > Do we want to have some kind of constraints in place for T ?
> 
> What kind of type constraints would you put on it?
I am not sure at this time but earlier I was thinking of interfaces like `Comparator`  but that still does add the readability that I was thinking of so I would say ignore my comment .

File: slices/slice_func.go [hunk 1]
```diff
@@ -191,6 +191,18 @@ func Map[S ~[]T, T any, G any](items S, f func(el T) G) []G {
 	return result
 }
 
+// Map applies F to all elements in slice of T and returns slice results as pointers, and will filter out the non-nil ones.
+func MapFilter[S ~[]T, T any, G any](items S, f func(el T) *G) []G {
```
Comment (orsinium) [inline slices/slice_func.go:195]: Maybe, using a separate bool argument would be more clean?

1. Returning the pointer can get messy when `G` itself is a pointer. Then you'd need to return `**int`.
2. Returning the bool would make it looks similar to how Filter works.

```suggestion
func MapFilter[S ~[]T, T any, G any](items S, f func(el T) (G, bool)) []G {
```

File: slices/slice_func_test.go [hunk 2]
```diff
@@ -164,6 +165,25 @@ func TestMap(t *testing.T) {
 	f([]int{1, 2, 3}, []int{2, 4, 6})
 }
 
+func TestMapFilter(t *testing.T) {
```
Comment (orsinium) [inline slices/slice_func_test.go:168]: Could you also add an `Example`, please? They are in `examples_test.go`. I think examples are very helpful for the end users.

**Comment by orsinium** [comment]
> Do we want to have some kind of constraints in place for T ?

No, I don't think it fits there. In this case, T can be of any type as long as the `f` function knows how to deal with it.

**Comment by calvinbrown085** [comment]
@orsinium I think I added all your suggestions, thanks for taking a look!

**Comment by orsinium** [comment]
Perfect! Thank you :)

**Pull request merged**